src/shared/buttercup/entries.js

```diff
@@ -50,77 +51,132 @@ export function validateEntry(entry) {
-
-        // Update properties
-        for (const propertyKey in properties) {
-          if (
-            properties.hasOwnProperty(propertyKey) && // eslint-disable-line no-prototype-builtins
-            entry.getProperty(propertyKey) !== properties[propertyKey]
-          ) {
-            entry.setProperty(propertyKey, properties[propertyKey]);
-          }
-        }
-
-        // Remove Meta
-        for (const metaKey in sourceMeta) {
-          if (sourceMeta.hasOwnProperty(metaKey)) {
-            // eslint-disable-line no-prototype-builtins
-            const keys = meta.map(metaObj => metaObj.key);
-            if (keys.indexOf(metaKey) === -1) {
-              entry.deleteMeta(metaKey);
-            }
-          }
-        }
-
-        // Update/Add meta
-        meta.forEach(metaObj => {
-          const source = entry.getMeta(metaObj.key);
-          if (
-            source === undefined ||
-            (source !== undefined && source !== metaObj.value)
-          ) {
-            entry.setMeta(metaObj.key, metaObj.value);
-          }
-        });
-
-        // Save workspace
-        saveWorkspace(archiveId);
-
-        resolve(entryToObj(entry));
-      })
-      .catch(err => reject(err));
+  const entryData = validateEntry(entryObj);
+
+  const properties = entryData.properties || {};
+  const meta = entryData.meta || [];
+  const sourceMeta = entry.toObject().meta || {};
+
+  // Update properties
+  for (const propertyKey in properties) {
+    if (
+      properties.hasOwnProperty(propertyKey) && // eslint-disable-line no-prototype-builtins
+      entry.getProperty(propertyKey) !== properties[propertyKey]
+    ) {
+      entry.setProperty(propertyKey, properties[propertyKey]);
+    }
+  }
+
+  // Remove Meta
+  for (const metaKey in sourceMeta) {
+    if (sourceMeta.hasOwnProperty(metaKey)) {
+      // eslint-disable-line no-prototype-builtins
+      const keys = meta.map(metaObj => metaObj.key);
+      if (keys.indexOf(metaKey) === -1) {
+        entry.deleteMeta(metaKey);
+      }
+    }
+  }
+
+  // Update/Add meta
+  meta.forEach(metaObj => {
+    const source = entry.getMeta(metaObj.key);
+    if (
+      source === undefined ||
+      (source !== undefined && source !== metaObj.value)
+    ) {
+      entry.setMeta(metaObj.key, metaObj.value);
+    }
   });
+
+  // Save workspace
+  saveWorkspace(archiveId);
+
+  return entryToObj(entry);
+}
+
+export async function updateEntryIcon(archiveId, entryId) {
+  const arch = getArchive(archiveId);
+  const entry = arch.getEntryByID(entryId);
+
+  if (!entry) {
+    throw new Error(i18n.t('error.entry-not-found'));
+  }
+
+  const entryObj = entryToObj(entry);
+  const icon = await fetchIcon(entry);
+  if (icon) {
+    entryObj.icon = icon;
+  }
+
+  return entryObj;
+}
+
+async function fetchIcon(entry) {
+  // We move on whether this succeeds or not
+  // TODO Should maybe log it somewhere (but not alert the user - not an error)
```

Comment:
Iconographer error flow maybe could be improved (we probably don't want to show Iconographer errors to the user, but it would be good to log them somewhere they can be at some point gathered)

Reply:
I would recommend using the [batched approach](https://github.com/buttercup/iconographer#managed-batch-icon-fetching) for the desktop app.. Immediate when seeing an icon for the first time, call `processIconForEntry ` and then for every entry just `getIconForEntry`. All `null`s should just default to the default icon.

Reply:
Basically, the default icon should always be visible and only replaced by a successfully fetched icon.

Reply:
@hugo-cardenas we have a module called `electron-log` which we are using to log stuff to user's userData directory

Reply:
@perry-mitchell I actually did not really get the point of having Iconographer both return a Promise and emitting Events. Probably I'm missing something.

Now I went down the rabbit hole and I still don't see it, maybe you can clarify this for me.

- So `Iconographer`.[processIconForURL](https://github.com/buttercup/iconographer/blob/master/source/Iconographer.js#L139) first enqueues a task fetching the wanted icon, and when this task is completed, it enqueues another task which stores the icon data using the storage.

- I see that `Channel`.[enqueue](https://github.com/buttercup/channel-queue/blob/master/source/Channel.js#L122) returns the task promise, and this `Task`.[queuedPromise](https://github.com/buttercup/channel-queue/blob/master/source/Task.js#L45) actually resolves when the task is [completed](https://github.com/buttercup/channel-queue/blob/master/source/Task.js#L98).

- So going back up the rabbit hole, the Promise returned by `Iconographer.processIconForURL` actually resolves when the icon has been downloaded and stored into disk.

From what I understand now, the event emission would be useful if I would want to react in some universal way anytime **any** icon gets fetched and stored by the Iconographer. This could be defined totally separate from the specific Iconographer usage.

But this is not the case, as whenever I would react to the icon being stored, it would mean I want to fetch it and update a very specific entry. This means that actually this reaction is totally tied to the Iconographer usage and to the context of a specific entry.

So in the end, using the EventEmitter instead of Promises basically replaces this code:
```javascript 
// Creating or updating entry here
// Dispatch already the basic create/update with the entry data
// Then: 
let icon = await iconographer.getIconForEntry(entry);
if (!icon) {
  await iconographer.processIconForEntry(entry);
  icon = await iconographer.getIconForEntry(entry);
}
// Set icon to entry and dispatch entry update
```

with this code:

```javascript 
// Creating or updating entry here
iconographer.once('iconAdded', async url => {
  if (url === entry.url) {
    const icon = await iconographer.getIconForURL(targetURL);
    // Set icon to entry and dispatch update
  }
});
iconographer.processIconForEntry(entry);
// We can dispatch the basic create/update with the entry data here, 
// as we are not waiting for the icon update
```
...that was a long write. I hope you understand my doubts 😅

Reply:
Yes I understand your point.. I think it's more clear the way you had it actually.  The event emitter I feel is still strong for some use cases, and I may use it in the mobile, but your use of the promise API is succinct. Let's stick with that!